[TRCL-570][feat] introduce startScan Event to synchronize CCD with e-beam during SPARC acq

pieleric opened the pull request:
From the moment that the acquisition ask the e-beam scanner to start and the e-beam is actually at the right position, it takes *some* time.  It's in the order of a few ms. So far, the only e-beam scanner was  semcomedi, and it was (surprisingly) constant in terms of time. It would  take between 2 and 5 ms. So it was good enough to just always wait 5ms.

We now also have semnidaq, and the NI-DAQ library introduces a lot longer latency. While it can be as "fast" as 10ms, it sometimes takes 40ms to start a scan. There are probably ways to make it a little faster, but it's  unlikely it will be possible to always keep all scanners < 5ms.
    
=> Use a new Event, startScan, for the e-beam scanner to report the e-beam is ready. Then, we can use it to directly start a CCD acquisition.

pieleric committed:
[doc] stream sync improve comments and logs

pieleric committed:
[clean-up] driver semcomedi: remove newPixel event support

No code ever used this functionality. The semnidaq has something a
little similar, but it's a hardware event. So it actually could even
brinkg confusion. Now that we'll add another Event, remove this code.

pieleric committed:
[feat] driver: add startScan Event to e-beam scanners

Introduce a new (and almost first) software Event on the e-beam
scanners: startScan. It is emitted at the very beginning of the first
frame of a scan. If a scan acquires multiple frames, startScan is
emitted just once.

This will be used for better software synchronization of SPARC
acquisitions.

pieleric committed:
[feat] acq SEMCCDMDStream use startScan Event to synchronize CCD with e-beam

From the momement that the acquisition ask the e-beam scanner to start
and the e-beam is actually at the right position, it takes *some* time.
It's in the order of a few ms. So far, the only e-beam scanner was
semcomedi, and it was (surprisingly) constant in terms of time. It would
take between 2 and 5 ms. So it was good enough to just always wait 5ms.

We now also have semnidaq, and the NI-DAQ library introduces a lot longer latency.
While it can be as "fast" as 10ms, it sometimes takes 40ms to start a
scan. There are probably ways to make it a little faster, but it's
unlikely it will be possible to always keep all scanners < 5ms.

=> Use a new Event, startScan, for the e-beam scanner to report the
e-beam is ready. Then, we can use it to directly start a CCD
acquisition.

K4rishma commented on src/odemis/driver/test/semnidaq_test.py:
```diff
@@ -860,6 +888,10 @@ def test_flow(self):
+        self.assertEqual(evt_counter.count, 1)
+        self.scanner.startScan.unsubscribe(evt_counter)
+
+
```
to remove the extra space between the functions

K4rishma commented on src/odemis/driver/simsem.py:
```diff
@@ -540,6 +544,9 @@ def _acquire_thread(self, callback):
                 self.data._waitSync()
+                if first_frame:
+                    self.parent._scanner.startScan.notify()
+                    first_frame = False
```
Does this automatically stop notifying something which is similar to the unsubscribe used in the testcases when self._acquisition_must_stop.clear() is called in finally: section?

pieleric replied:
No, this is different. It's a model.Event, an Odemis object that works in between Components. `_acquisition_must_stop` is a threading.Event, a standard Python object that only works within the component (because each component runs in a separate Python process).

tmoerkerken commented on src/odemis/acq/stream/_sync.py:
```diff
@@ -1887,26 +1888,10 @@ def _acquireImage(self, n, px_idx, img_time, sem_time,
                 raise CancelledError()
 
             # subscribe to _subscribers
+            # As soon as the e-beam start scanning (which can take a couple of ms), the
```
```suggestion
            # As soon as the e-beam starts scanning (which can take a couple of ms), the
```

tmoerkerken commented on src/odemis/acq/stream/_sync.py:
```diff
@@ -1887,26 +1888,10 @@ def _acquireImage(self, n, px_idx, img_time, sem_time,
             # subscribe to _subscribers
+            # As soon as the e-beam start scanning (which can take a couple of ms), the
+            # startScan event is sent, which triggers the acquisition of one CCD frame.
             for s, sub in zip(self._streams[:-1], self._subscribers[:-1]):
```
Can you comment on why the last element is excluded?

tmoerkerken commented on src/odemis/acq/stream/_sync.py:
```diff
@@ -2119,12 +2112,11 @@ def _runAcquisitionScanStage(self, future):
                     if self._acq_state == CANCELLED:
                         raise CancelledError()
 
+                    # Start e-beam scan. As soon as it really start, a startScan event is sent, which
```
```suggestion
                    # Start e-beam scan. As soon as it really starts, a startScan event is sent, which
```

tmoerkerken commented on src/odemis/acq/stream/_sync.py:
```diff
@@ -2008,9 +2001,9 @@ def _runAcquisitionScanStage(self, future):
         # The idea of the acquiring with a scan stage:
```
```suggestion
        # The idea of acquisition with a scan stage:
```

tmoerkerken commented on src/odemis/driver/semnidaq.py:
```diff
@@ -2251,6 +2253,23 @@ def _set_state(self, active):
                 time.sleep(self._activation_delay)
 
 
+class EventOnce(model.Event):
```
This is duplication of the function at the semcomedi driver. Maybe put it in a shared module?

pieleric replied:
Indeed. moved to `driver.EventOnce`

K4rishma commented on src/odemis/acq/stream/_sync.py:
```diff
@@ -2210,6 +2202,11 @@ def _runAcquisitionScanStage(self, future):
+                                # Temporarily switch the CCD to a different event trigger, so that it
+                                # doesn't get triggered while the leech is running (because it could use the
+                                # e-beam, which would send a startScan event)
+                                self._ccd_df.synchronizedOn(self._ccd.softwareTrigger)
```
Maybe an assert line can be used to check that the notification boolean value of startscan is False before reusing the real trigger? Same for _acquireImage

pieleric committed:
[feat] new driver.EventOnce helper class for model.Event

In some cases, it can simplify the code if a notification should be sent
only of the first type of event. => introduce EventOnce.

The SEM DAQ board drivers will use it to implement the startScan event.
